### PR TITLE
Use reflection to get cache layer from J9shrOffset

### LINKED ISSUE #6971: 32-bit doesn't compile SharedClassesMetaDataHelper.java
https://ci.eclipse.org/openj9/job/Build_JDK8_x86-32_windows_OMR/264
```
22:02:04  c:\Users\jenkins\workspace\Build_JDK8_x86-32_windows_OMR\openj9\debugtools\DDR_VM\src\com\ibm\j9ddr\vm29\pointer\helper\SharedClassesMetaDataHelper.java:31: error: cannot find symbol
22:02:04  			layer = j9shrOffset.cacheLayer().intValue();
22:02:04  			                   ^
22:02:04    symbol:   method cacheLayer()
22:02:04    location: variable j9shrOffset of type J9ShrOffsetPointer
```

Caused by https://github.com/eclipse/openj9/pull/6883

### PR DESCRIPTION
**Pull request by hangshao0**
Fixes #6971

Signed-off-by: Hang Shao <hangshao@ca.ibm.com>

### CODE REVIEW

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java [hunk 1]
```diff
@@ -23,14 +23,20 @@
...
+import com.ibm.j9ddr.vm29.types.UDATA;
+import java.lang.reflect.*;
 
 public class SharedClassesMetaDataHelper {
 	public static int getCacheLayerFromJ9shrOffset(J9ShrOffsetPointer j9shrOffset) throws CorruptDataException {
 		int layer;
 		try {
-			layer = j9shrOffset.cacheLayer().intValue();
-		} catch (NoSuchFieldError | NoSuchMethodError e) {
+			Method m = j9shrOffset.getClass().getMethod("cacheLayer");
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java]: Is this method called frequently enough that it should cache the method?

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java [hunk 2]
```diff
@@ -23,14 +23,20 @@
...
+			UDATA layerUDATA = (UDATA) m.invoke(j9shrOffset);
+			layer = layerUDATA.intValue();
+		} catch (NoSuchMethodException | InvocationTargetException e) {
 			layer = 0;
+		} catch (Throwable e) {
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java]: It's generally a bad idea to catch `Throwable`: what exceptions are anticipated?

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java [hunk 3]
```diff
@@ -23,14 +23,20 @@
 
 import com.ibm.j9ddr.CorruptDataException;
 import com.ibm.j9ddr.vm29.pointer.generated.J9ShrOffsetPointer;
+import com.ibm.j9ddr.vm29.types.UDATA;
+import java.lang.reflect.*;
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java]: I believe this only imports one class (Method); a single import is preferable to a wildcard in that case.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java (hunk 1 above)
Comment (hangshao0) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java]: Remove `Method m`.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java (hunk 3 above)
Comment (hangshao0) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java]: Fixed

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java (hunk 2 above)
Comment (hangshao0) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java]: Removed this block. I expect `NoSuchMethodException, InvocationTargetException`. But compiler complains about `IllegalAccessException` from `Method.invoke()`, so I've added IllegalAccessException into the exception list in the previous block.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java [hunk 4]
```diff
@@ -23,13 +23,16 @@
...
 
 public class SharedClassesMetaDataHelper {
 	public static int getCacheLayerFromJ9shrOffset(J9ShrOffsetPointer j9shrOffset) throws CorruptDataException {
 		int layer;
 		try {
-			layer = j9shrOffset.cacheLayer().intValue();
-		} catch (NoSuchFieldError | NoSuchMethodError e) {
+			UDATA layerUDATA = (UDATA) j9shrOffset.getClass().getMethod("cacheLayer").invoke(j9shrOffset);
+			layer = layerUDATA.intValue();
+		} catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java]: In the case of `InvocationTargetException`, I think it should throw the 'cause', if it is a `CorruptDataException` or `RuntimeException`.

Comment (hangshao0) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java]: Change has been updated.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java [hunk 5]
```diff
@@ -23,14 +23,26 @@
...
-		} catch (NoSuchFieldError | NoSuchMethodError e) {
+			UDATA layerUDATA = (UDATA) j9shrOffset.getClass().getMethod("cacheLayer").invoke(j9shrOffset);
+			layer = layerUDATA.intValue();
+		} catch (NoSuchMethodException e) {
 			layer = 0;
+		} catch (InvocationTargetException e) {
+			Throwable cause = e.getCause();
+			if (cause instanceof NoSuchFieldError) {
+				layer = 0;
+			} else {
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java]: My suggestion was to include checks for `CorruptDataException` and `RuntimeException`; i.e. insert the following before this line:
```
else if (cause instanceof CorruptDataException) {
    throw (CorruptDataException) cause;
}
else if (cause instanceof RuntimeException) {
    throw (RuntimeException) cause;
```

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java [hunk 6]
```diff
@@ -23,14 +23,26 @@
...
+				throw new CorruptDataException("Error getting the cache layer from J9ShrOffsetPointer");
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java]: Please chain the cause into this exception.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java (hunk 5 above)
Comment (hangshao0) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java]: I see. Updated.

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java [hunk 7]
```diff
@@ -23,14 +23,30 @@
...
+		} catch (InvocationTargetException e) {
+			Throwable cause = e.getCause();
+			if (cause instanceof NoSuchFieldError) {
+				layer = 0;
+			} else if (cause instanceof CorruptDataException) {
+				throw (CorruptDataException) cause;
+			} else if (cause instanceof RuntimeException) {
+				throw (RuntimeException) cause;
+			} else {
+				throw new CorruptDataException("Error getting the cache layer from J9ShrOffsetPointer: " + e.getCause());
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java]: I meant you should use the two-argument constructor: `CorruptDataException(String message, Throwable t)`.

Comment (hangshao0) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/pointer/helper/SharedClassesMetaDataHelper.java]: Done

**Comment by keithc-ca** [comment]
Jenkins test sanity win,win32 jdk8

**Pull request merged**